### PR TITLE
CKS: create Security Groups for CKS clusters of each account

### LINKED ISSUE #7684: CKS in Advanced Zone with SG: Avoid adding new Security Groups
##### ISSUE TYPE
 * Improvement Request

##### COMPONENT NAME
~~~
Kubernetes
~~~

##### CLOUDSTACK VERSION
~~~
4.17.2
~~~

##### CONFIGURATION
~~~
Advanced Zone with SG
~~~

##### SUMMARY
Each time when a user creates a KCS in an Advanced Zone with SG a new Security Group is added:
image
This results in having a vast amount of Security Groups where a single Security Group would be sufficient. I propose adding a Dropdown List in the KCS creation view for choosing a SG, similar to the Add Instance UI.

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

This PR fixes #7684 

The security groups contain the same rules for port 22 and 6443, no need to recreate for each CKS cluster.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Review by Pearl1594 — APPROVED** [review]
Code LGTM

**Comment by weizhouapache** [comment]
@shwstppr @DaanHoogland 
this is a simple PR,  we could consider it in 4.19.0.0

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@DaanHoogland 
the advanced zone should not be impacted by this change.
may need to test advanced zone with SG manually

**Comment by DaanHoogland** [comment]
@blueorangutan test alma9 kvm-alma9 securityGroups keepEnv

File: server/src/main/java/com/cloud/network/security/SecurityGroupManagerImpl.java [hunk 1]
```diff
@@ -1410,7 +1407,7 @@ public SecurityGroupVO getDefaultSecurityGroup(long accountId) {
     }
 
     @Override
-    public SecurityGroup getSecurityGroup(String name, long accountId) {
+    public SecurityGroupVO getSecurityGroup(String name, long accountId) {
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/network/security/SecurityGroupManagerImpl.java:1410]: @weizhouapache, would it be better to still return `SecurityGroup` in the calling method and maybe add methods to the interface? We are adding coupling this way.

Comment (weizhouapache) [reply server/src/main/java/com/cloud/network/security/SecurityGroupManagerImpl.java]: done

**Comment by weizhouapache** [comment]
@blueorangutan package

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterManagerImpl.java [hunk 2]
```diff
@@ -1068,30 +1068,35 @@ public KubernetesCluster createKubernetesCluster(CreateKubernetesClusterCmd cmd)
             logAndThrow(Level.ERROR, String.format("Creating Kubernetes cluster failed due to error while finding suitable deployment plan for cluster in zone : %s", zone.getName()));
         }
 
-        SecurityGroupVO securityGroupVO = null;
+        SecurityGroup securityGroupVO = null;
         if (zone.isSecurityGroupEnabled()) {
```
Comment (DaanHoogland) [inline plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterManagerImpl.java:1072]: can this block be extracted to a separate method?

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterManagerImpl.java [hunk 3]
```diff
@@ -1068,30 +1068,35 @@ public KubernetesCluster createKubernetesCluster(CreateKubernetesClusterCmd cmd)
             logAndThrow(Level.ERROR, String.format("Creating Kubernetes cluster failed due to error while finding suitable deployment plan for cluster in zone : %s", zone.getName()));
         }
 
-        SecurityGroupVO securityGroupVO = null;
+        SecurityGroup securityGroupVO = null;
```
Comment (DaanHoogland) [inline plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterManagerImpl.java:1071]: `securityGroupVO` could also be shortened to `securityGroup`.

**Review by DaanHoogland — APPROVED** [review]
tested by creating two k8 clusters as a user and a third as another user.
two security groups and sets of rules are created, the first shared amongst the first user's clusters.
LGTM (but for some style remarks)

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterManagerImpl.java (hunk 2 above)
Comment (weizhouapache) [reply plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterManagerImpl.java:1072]: done @DaanHoogland

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterManagerImpl.java (hunk 3 above)
Comment (weizhouapache) [reply plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterManagerImpl.java]: done

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Pull request merged**